### PR TITLE
[canary] Refactor: `PromptedAudioToTextLhotseDataset` and `EncDecMultiTaskModel`

### PR DESCRIPTION
**Pull request by pzelasko**
This PR restores `transformer_bpe_models.py` and `audio_to_text_lhotse.py` to their previous shape and separates out two classes: `EncDecMultiTaskModel` and `PromptedAudioToTextLhotseDataset` that contain the modified code for Canary, generalized to support other model instances of this type.

### CODE REVIEW

File: nemo/collections/asr/data/audio_to_text_canary.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+from lhotse.cut import MixedCut, MonoCut
+from lhotse.dataset import AudioSamples
+from lhotse.dataset.collation import collate_vectors
+
+from nemo.collections.asr.data.audio_to_text_lhotse import TokenizerWrapper
+
+
+class CanaryDataset(torch.utils.data.Dataset):
+    """
+    This dataset is based on :class:`~nemo.collections.asr.data.audio_to_text_lhotse.LhotseSpeechToTextBpeDataset`.
```
Comment (zhehuaichen) [inline nemo/collections/asr/data/audio_to_text_canary.py:25]: why not make it inherit LhotseSpeechToTextBpeDataset to easier show the differences?

File: nemo/collections/asr/data/audio_to_text_canary.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    def __init__(self, tokenizer):
+        super().__init__()
+        self.tokenizer = TokenizerWrapper(tokenizer)
+        self.load_audio = AudioSamples(fault_tolerant=True)
+        self.padding_value = self.tokenizer._tokenizer.pad_id
+
+    def __getitem__(self, cuts) -> tuple[torch.Tensor, ...]:
+        audio, audio_lens, cuts = self.load_audio(cuts)
+
+        tokens = [self.tokenizer(c.supervisions[0].text, c.supervisions[0].language) for c in cuts]
```
Comment (zhehuaichen) [inline nemo/collections/asr/data/audio_to_text_canary.py:40]: what do you think about merging c.supervisions[0].language with c.custom['target_lang'] to reduce redundancy and confusion?

**Review by zhehuaichen — COMMENTED** [review]
Thanks for the work!

File: nemo/collections/asr/data/audio_to_text_canary.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        tokens = self._canary_format(tokens, cuts)
+        tokens = [torch.as_tensor(t) for t in tokens]
+        token_lens = torch.tensor([t.size(0) for t in tokens], dtype=torch.long)
+        tokens = collate_vectors(tokens, padding_value=self.padding_value)
+
+        return audio, audio_lens, tokens, token_lens
+
+    def _canary_format(self, tokens, cuts):
```
Comment (stevehuang52) [inline nemo/collections/asr/data/audio_to_text_canary.py:48]: Could you please add type annotations to the params for `__getitem__()` and `_canary_format()`?

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 4]
```diff
@@ -367,9 +367,7 @@ def _setup_dataloader_from_config(self, config: Optional[Dict]):
                 config,
                 global_rank=self.global_rank,
                 world_size=self.world_size,
-                dataset=LhotseSpeechToTextBpeDataset(
-                    tokenizer=self.tokenizer, token_sequence_format=config.get("token_sequence_format", None),
-                ),
+                dataset=CanaryDataset(tokenizer=self.tokenizer),
```
Comment (stevehuang52) [inline nemo/collections/asr/models/transformer_bpe_models.py:370]: This will force dataset to `CanaryDataset` but sometimes users might want simply `LhotseSpeechToTextBpeDataset` instead. How about create a `Canary` model class which inherits  the `EncDecTransfModelBPE` class? Given that Canary is doing so well, I think it deserves its own model class, which will make it easier for use to update later without worrying about changing the behaviors of existing `EncDecTransfModelBPE` models. What do you think?

File: nemo/collections/asr/data/audio_to_text_canary.py (hunk 1 above)
Comment (pzelasko) [reply nemo/collections/asr/data/audio_to_text_canary.py]: We settled on `PromptedAudioToTextLhotseDataset` to support more similar use-cases in the future.

File: nemo/collections/asr/data/audio_to_text_canary.py (hunk 2 above)
Comment (pzelasko) [reply nemo/collections/asr/data/audio_to_text_canary.py]: Hmmm we could technically remove `text_field` and `lang_field` from the custom fields when constructing cuts from NeMo JSON manifest, but I think I'd rather keep it for easy traceability... WDYT?

File: nemo/collections/asr/models/transformer_bpe_models.py (hunk 4 above)
Comment (pzelasko) [reply nemo/collections/asr/models/transformer_bpe_models.py]: Done

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+__all__ = ['EncDecMultiTaskModel']
+
+
+def lens_to_mask(lens, max_length):
+    batch_size = lens.shape[0]
+    mask = torch.arange(max_length).repeat(batch_size, 1).to(lens.device) < lens[:, None]
+    return mask
+
+
+class EncDecMultiTaskModel(ASRModel, ExportableEncDecModel, ASRBPEMixin):
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:78]: TODO @titu1994  Add ASRAdapterMixin, ASRModuleMixin, TranscriptionMixin after code is stable.

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+        if self.cfg.encoder['d_model'] != self.cfg.transf_decoder['hidden_size']:
+            self.adapter = torch.nn.Linear(self.cfg.encoder['d_model'], self.cfg.transf_decoder['hidden_size'])
+        else:
+            self.adapter = torch.nn.Identity()
+
+        transf_encoder_cfg_dict = OmegaConf.to_container(cfg.get('transf_encoder'))
+
+        # Whether to add Transformer Encoder block between Conformer and Transformer Decoder
+        self.use_transf_encoder = False
+        if transf_encoder_cfg_dict['num_layers'] > 0:
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:107]: Revert  manual class instantiation and use `from_config_dict` from above.

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
+
+        transf_decoder_cfg_dict = OmegaConf.to_container(cfg.get('transf_decoder'))
+
+        # Transformer decoder
+        vocab_size = 8 * ceil(self.tokenizer.vocab_size / 8)
+        transf_decoder_cfg_dict['vocab_size'] = vocab_size
+        library = transf_decoder_cfg_dict.pop('library', 'nemo')
+        model_name = transf_decoder_cfg_dict.pop('model_name', None)
+        pretrained = transf_decoder_cfg_dict.pop('pretrained', False)
+        self.transf_decoder = get_transformer(
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:133]: Instantiate via config

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 8]
```diff
@@ -0,0 +1,726 @@
...
+            library=library,
+            model_name=model_name,
+            pretrained=pretrained,
+            config_dict=transf_decoder_cfg_dict,
+            encoder=False,
+            pre_ln_final_layer_norm=transf_decoder_cfg_dict.get("pre_ln_final_layer_norm", False),
+        )
+
+        self.log_softmax = TokenClassifier(
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:142]: Instantiate via Config

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 9]
```diff
@@ -0,0 +1,726 @@
...
+            dropout=self.cfg.head.dropout,
+            use_transformer_init=self.cfg.head.use_transformer_init,
+        )
+        self.log_softmax.mlp.layer0.weight = self.transf_decoder.embedding.token_embedding.weight
+        std_init_range = 1 / self.transf_decoder.hidden_size ** 0.5
+        self.transf_decoder.apply(lambda module: transformer_weights_init(module, std_init_range))
+        self.log_softmax.apply(lambda module: transformer_weights_init(module, std_init_range))
+
+        # Beam Search decoding
+        self.beam_search = BeamSearchSequenceGenerator(
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:156]: TODO @titu1994 Move this to Decoding strategy

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 10]
```diff
@@ -0,0 +1,726 @@
...
+            pad=self.tokenizer.pad_id,
+            eos=self.tokenizer.eos_id,
+            len_pen=self.cfg.beam_search.len_pen,
+            max_delta_length=self.cfg.beam_search.max_generation_delta,
+        )
+        # TO DO: remove this hardcoded context size for AR decoding
+        self.context_len_for_AR_decoding = 5
+
+        # Define autoregressive CE loss
+        self.transf_loss = SmoothedCrossEntropyLoss(
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:172]: Instantiate via Config

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 11]
```diff
@@ -0,0 +1,726 @@
...
+            self.spec_augmentation = EncDecMultiTaskModel.from_config_dict(self.cfg.spec_augment)
+        else:
+            self.spec_augmentation = None
+
+        self.val_loss = GlobalAverageLossMetric(dist_sync_on_step=False, take_avg_loss=True)
+
+    def change_decoding_strategy(self, cfg: DictConfig):
+        logging.info(f"Changing beam search decoding to {cfg}")
+        # Beam Search decoding
+        self.beam_search = BeamSearchSequenceGenerator(
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:186]: TODO @titu1994  Move to Decoding strategy

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 12]
```diff
@@ -0,0 +1,726 @@
...
+            return_hypotheses: (bool) Either return hypotheses or text
+                With hypotheses can do some postprocessing like getting timestamp or rescoring
+            num_workers: (int) number of workers for DataLoader
+            channel_selector (int | Iterable[int] | str): select a single channel or a subset of channels from multi-channel audio. If set to `'average'`, it performs averaging across channels. Disabled if set to `None`. Defaults to `None`.
+            augmentor: (DictConfig): Augment audio samples during transcription if augmentor is applied.
+            verbose: (bool) whether to display tqdm progress bar
+        Returns:
+            A list of transcriptions (or raw log probabilities if logprobs is True) in the same order as paths2audio_files
+        """
+        if paths2audio_files is None or len(paths2audio_files) == 0:
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:229]: TODO @titu1994 Move to Transcription Mixin later

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 13]
```diff
@@ -0,0 +1,726 @@
...
+            pad=self.tokenizer.pad_id,
+            eos=self.tokenizer.eos_id,
+            len_pen=cfg.len_pen,
+            max_delta_length=cfg.max_generation_delta,
+        )
+
+    @torch.no_grad()
+    def transcribe(
+        self,
+        paths2audio_files: Union[List[str], str],
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:202]: Make this key `audio`

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 14]
```diff
@@ -0,0 +1,726 @@
...
+        batch_size: int = 4,
+        logprobs: bool = False,
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:204]: Remove logprobs , its deprecated

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 15]
```diff
@@ -0,0 +1,726 @@
...
+
+        beam_hypotheses = self.beam_search(
+            encoder_hidden_states=enc_states,
+            encoder_input_mask=enc_mask,
+            return_beam_scores=False,
+            decoder_input_ids=input_ids[:, : self.context_len_for_AR_decoding]
+            if self.context_len_for_AR_decoding > 0
+            else None,
+        )
+        transf_loss = self.transf_loss(log_probs=transf_log_probs, labels=labels)
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:601]: Reorder steps to do forward -> loss -> beam -> metric -> return dict

File: nemo/collections/asr/models/aed_multitask_models.py [hunk 16]
```diff
@@ -0,0 +1,726 @@
...
+
+            # in multi-validation case, anything after first one will become NaN
+            # as we are resetting the metric here.
+            # TODO: fix this, (not sure which hook will be ideal for this)
+            self.val_loss.reset()
+
+    def multi_test_epoch_end(self, outputs, dataloader_idx: int = 0):
+        return self.multi_validation_epoch_end(outputs, dataloader_idx, eval_mode="test")
+
+    def test_dataloader(self):
```
Comment (titu1994) [inline nemo/collections/asr/models/aed_multitask_models.py:691]: Can be removed?

File: nemo/collections/asr/models/transformer_bpe_models.py [hunk 17]
```diff
@@ -178,26 +173,10 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
 
         self.val_loss = GlobalAverageLossMetric(dist_sync_on_step=False, take_avg_loss=True)
 
-    def change_decoding_strategy(self, cfg: DictConfig):
```
Comment (titu1994) [inline nemo/collections/asr/models/transformer_bpe_models.py:181]: @AlexGrinch  this can be kept in another PR

**Review by titu1994 — COMMENTED** [review]
Loads of comments, please go through them and lets discuss if something is impossible to do right now

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+    include: ["preprocessor", "encoder"]
+
+model:
+  _target_: nemo.collections.asr.models.EncDecMultiTaskModel
+  sample_rate: 16000
+  label_smoothing: 0.0
+  log_prediction: true # enables logging sample predictions in the output during training
+
+  train_ds:
+    is_tarred: true
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:25]: lhotse will ignore it anyway, I mised this -- removing

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    tarred_audio_filepaths: ???
+    manifest_filepath: ???
+    sample_rate: 16000
+    shuffle: true
+    trim_silence: false
+    num_workers: 8
+    use_lhotse: true
+    batch_size: None
+    batch_duration: 360
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:34]: I'll add a comment with a link to Lhotse docs where this is explained

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+  label_smoothing: 0.0
+  log_prediction: true # enables logging sample predictions in the output during training
+
+  train_ds:
+    is_tarred: true
+    tarred_audio_filepaths: ???
+    manifest_filepath: ???
+    sample_rate: 16000
+    shuffle: true
+    trim_silence: false
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:30]: Removed, Lhotse ignores it anyway

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+    pin_memory: true
+    use_start_end_token: true
+    use_lhotse: true
+    use_bucketing: false
+
+  # recommend small vocab size of 128 or 256 when using 4x sub-sampling
+  # you may find more detail on how to train a tokenizer at: /scripts/tokenizers/process_asr_text_tokenizer.py
+  tokenizer:
+    dir: ???  # path to directory which contains either tokenizer.model (bpe) or vocab.txt (wpe)
+    type: bpe  # Can be either bpe (SentencePiece tokenizer) or wpe (WordPiece tokenizer)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:67]: Let's add in `canary-2`

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+    conv_norm_type: batch_norm
+    conv_context_size: null
+
+    ### regularization
+    dropout: 0.1 # The dropout used in most of the Conformer Modules
+    dropout_pre_encoder: 0.1
+    dropout_emb: 0.0 # The dropout used for embeddings
+    dropout_att: 0.1 # The dropout for multi-headed attention modules
+
+  transf_encoder:
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:131]: I think this is the unused "extra" transformer encoder between the actual encoder and the decoder that @AlexGrinch mentioned during the presentation

File: nemo/collections/asr/data/audio_to_text_lhotse_prompted.py [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+    return canary_tokens
+
+
+# Mapping from a string name to a known prompt formatter function.
+PROMPT_FORMAT_FNS = {
+    "canary": _canary_prompt_format,
+}
+
+
+def get_prompt_format_fn(name: str) -> Callable[[Sequence[Sequence[int]], CutSet], Sequence[Sequence[int]]]:
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/data/audio_to_text_lhotse_prompted.py]: Done

**Comment by pzelasko** [comment]
Merging to `canary-2` for further development to unblock others -- we should address the rest of comments there.

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml (hunk 20 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml]: Ah but the default in our code is True by default :headwall:

File: examples/asr/conf/speech_multitask/fast-conformer_aed.yaml (hunk 22 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply examples/asr/conf/speech_multitask/fast-conformer_aed.yaml:135]: Huh? Do our current models need it? If not throw it out. We can create a script to drop the unused parameters out of current models.
